cairo/crates/contracts/src/hooks/interchain_gas_paymaster.cairo

```diff
@@ -0,0 +1,404 @@
+
+            let caller = get_caller_address();
+            let gas_token = ERC20ABIDispatcher { contract_address: ETH_ADDRESS() };
+            gas_token.transfer_from(caller, get_contract_address(), _payment);
+
+            let refund_amount = _payment - required;
+            if refund_amount > 0 {
+                gas_token.transfer( _refund_address, refund_amount);
+                self.emit(GasRefund { refund_address: _refund_address, amount: refund_amount });
+            }
+
+            self.emit(GasPayment { message_id: _message_id, destination_domain: _destination_domain, gas_amount: _gas_amount, payment: required });
+        }
+
+        /// Quotes gas payment amount for a given destination domain and gas amount.
+        ///
+        /// # Arguments
+        ///
+        /// * `_destination_domain` - Domain ID of the destination chain
+        /// * `_gas_amount` - Amount of gas to quote for on the destination chain
+        ///
+        /// # Returns
+        ///
+        /// u256 - Total required payment in native tokens
+        fn quote_gas_payment(
+            ref self: ContractState,
+            _destination_domain: u32,
+            _gas_amount: u256,
+        ) -> u256 {
+            let config = self.gas_configs.read(_destination_domain);
+            assert(config.token_exchange_rate != 0, Errors::CONFIG_NOT_FOUND);
+            let total_gas: u256 = _gas_amount + config.gas_overhead;
+            let dest_cost: u256 = total_gas * config.gas_price.into();
+            (dest_cost * config.token_exchange_rate.into()) / TOKEN_EXCHANGE_RATE_SCALE
+        }
+
+        /// Sets gas configuration for multiple destination domains.
+        ///
+        /// # Arguments
+        ///
+        /// * `configs` - Array of GasParam configurations to set
+        fn set_destination_gas_configs(
+            ref self: ContractState,
+            configs: Array<GasParam>
+        ){
+            self.ownable.assert_only_owner();
+            for config in configs{
+                self._set_destination_gas_config(config.remote_domain, config.config);
+            }
+        }
+
+        /// Sets the beneficiary address that can withdraw collected payments.
+        ///
+        /// # Arguments
+        ///
+        /// * `beneficiary` - New beneficiary address
+        fn set_beneficiary(
+            ref self: ContractState,
+            beneficiary: ContractAddress
+        ) {
+            self.ownable.assert_only_owner();
+            self._set_beneficiary(beneficiary)
+        }
+
+        /// Allows the beneficiary to claim all collected ETH tokens.
+        ///
+        /// # Returns
+        ///
+        /// u256 - Amount claimed
+        fn claim(ref self: ContractState) -> u256 {
+            let caller = get_caller_address();
+            let beneficiary = self.beneficiary.read();
+            
+            assert(caller == beneficiary, Errors::NOT_BENEFICIARY);
```

Comment:
we should allow anyone to call this function but the funds only get transferred to the beneficiary